crypto/fipsmodule/rand/rand.c

```diff
@@ -302,51 +340,64 @@ static void rand_get_seed(struct rand_thread_state *state,
 // rand_get_seed fills |seed| with entropy and sets |*out_used_cpu| to one if
 // that entropy came directly from the CPU and zero otherwise.
 static void rand_get_seed(struct rand_thread_state *state,
-                          uint8_t seed[CTR_DRBG_ENTROPY_LEN],
-                          int *out_used_cpu) {
+                          uint8_t seed[CTR_DRBG_MAX_ENTROPY_LEN],
+                          int *out_used_cpu,
+                          size_t entropy_len) {
+
   // If not in FIPS mode, we don't overread from the system entropy source and
   // we don't depend only on the hardware RDRAND.
-  CRYPTO_sysrand(seed, CTR_DRBG_ENTROPY_LEN);
+  CRYPTO_sysrand(seed, entropy_len);
   *out_used_cpu = 0;
 }
 
 #endif
 
+// We want to ensure that |CTR_DRBG_MAX_AES_KEY_LEN| is always 32, because this
+// is assumed higher up the stack e.g. big number library module (see
+// |bn_rand_range_words()|). But we do not want to pollute the rest of the code
+// with includes and |CTR_DRBG_MAX_AES_KEY_LEN|. The assert below will catch
+// if the max value ever changes.
+OPENSSL_STATIC_ASSERT(CTR_DRBG_MAX_AES_KEY_LEN == 32,
+  CTR_DRBG_AES_max_key_size_is_not_32)
+
 void RAND_bytes_with_additional_data(uint8_t *out, size_t out_len,
-                                     const uint8_t user_additional_data[32]) {
+                                     const uint8_t user_additional_data[CTR_DRBG_MAX_AES_KEY_LEN],
+                                     ctr_drbg_key_len_t ctr_drbg_key_len) {
   if (out_len == 0) {
     return;
   }
 
+  size_t entropy_len = ctr_drbg_get_entropy_length(ctr_drbg_key_len);
+  size_t aes_key_len = ctr_drbg_get_key_length(ctr_drbg_key_len);
+
   const uint64_t fork_generation = CRYPTO_get_fork_generation();
 
   // Additional data is mixed into every CTR-DRBG call to protect, as best we
   // can, against forks & VM clones. We do not over-read this information and
   // don't reseed with it so, from the point of view of FIPS, this doesn't
   // provide “prediction resistance”. But, in practice, it does.
-  uint8_t additional_data[32];
```

Comment:
So as per table 3, it can be up to _seedlen = outlen + keylen_, right?

Reply:
That's right. Correctness is not trivial to deduce. Further details below.

10.2.1.5.1:
* `CTR_DRBG Generate Process` specifies that `additional_input` must be zero-padded and does this explicitly in step 2.2.
* After this step, the update function `CTR_DRBG_UPDATE` is called with the padded `additional_input`.

The AWS-LC implementation is not 1-1 with this, but is functionally equivalent. To see why, we need to go to the AWS-LC update function `ctr_drbg_update`.
* The update function (spec cross-reference 10.2.1.2, step 4 of `CTR_DRBT_Update process`) XOR's the additional data (named `provided_data` in the spec) into the output stream of CTR-AES.
* The additional data is not part of producing the output stream.
* Since xor'ing with 0 doesn't do anything we simply needs to xor in the number of bytes in `additional_data`. This is done here (original code): https://github.com/awslabs/aws-lc/blob/main/crypto/fipsmodule/rand/ctrdrbg.c#L92-L94

The changes made in this PR still have the equivalent functionality: we only use `drbg->entropy_len` bytes of the temp array `tmp`. `additional_data_len` is always less or equal to `drbg->entropy_len`.

Reply:
Thank you for the detailed explanation, Torben.